Fixed InputManager.asset file growing in size on each Reset call

lyndon-unity opened the pull request:
### Description

Reset now clears out old Input Action Assets and Input Action References first to avoid the InputManager.asset file growing in size over each reset

For bug https://jira.unity3d.com/browse/ISX-1839 

### Changes made

Spinning through contents of InputManager.asset to remove Input Action Assets and Input Action References first 
Then using GetOrCreate to create a new version (to make sure only 1 added, if GetOrCreate called in a callback elsewhere - which it is due to AssetDatabase.SaveAssets call)

### Notes

### Checklist

Before review:

- [X] Changelog entry added.
- [ ] Tests added/changed, if applicable.
- None added on this instance, but opportunity to add one for this

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

lyndon-unity committed:
Reset now clears out old Input Action Assets and Input Action References first

lyndon-unity committed:
Added change log and formatting fix

ekcoh commented on Packages/com.unity.inputsystem/CHANGELOG.md:
```diff
@@ -34,6 +34,7 @@ however, it has to be formatted properly to pass verification tests.
 - Fixed issue where composite part dropdown manipulates binding path and leaves composite part field unchanged.
 - Fixed lingering highlight effect on Save Asset button after clicking.
 - Fixed missing name in window title for Input Action assets.
+- Fixed InputManager.asset file growing in size on each Reset call (ISX-1839)
```
Minor: Missing full-stop in end. Referencing internal case ID for a non-publicly visible issue.

lyndon-unity replied:
Changed (online in githib editor ;) )
Note: Some prior change log entries also seem to have some internal bug references but I've not changed those in this PR

ekcoh reviewed (approved):
LGTM, a minor comment about changelog.

ekcoh commented:
Seems like PR got hit by CI instability, rerun tests?

lyndon-unity committed:
Update CHANGELOG.md

Removed internal bug id and added full stop

lyndon-unity commented:
> Seems like PR got hit by CI instability, rerun tests?

Re-running with minor changelog update.

ritamerkl commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs:
```diff
@@ -147,7 +149,46 @@ internal static void CheckForDefaultUIActionMapChanges()
+        /// <summary>
+        /// Delete project wide input actions
+        /// </summary>
+        /// <param name="asset"></param>
```
parameter left over in comment

ritamerkl commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs:
```diff
@@ -147,7 +149,46 @@ internal static void CheckForDefaultUIActionMapChanges()
+        /// <summary>
+        /// Reset project wide input actions asset
+        /// </summary>
+        /// <param name="asset"></param>
```
unused param asset

ritamerkl reviewed (approved):
Just two little remarks on the comments, looks good, nice to have this fixed!

lyndon-unity committed:
Moved the DeleteActionAsset call into CreateNewActionAsset to make sure we never create duplicates

Added Start/EndAssetEditing calls to try and make the serialisation code more atomic
Removed some unused param fields in the docs

Removed some excessiver SaveAssets calls

lyndon-unity committed:
Merge branch 'ISX_1839_reset_fix_to_remove_duplicates' of github.com:Unity-Technologies/InputSystem into ISX_1839_reset_fix_to_remove_duplicates

lyndon-unity committed:
Removed StartAssetEditing at AssetDatabase team recommendation as it doesn't actually impact AddObjectToAsset

lyndon-unity commented:
Params documentation fixed - and moved delete into the create so that when new data created any old data is always cleaned out. This catches an extra case.

lyndon-unity committed:
Formatting fix

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs:
```diff
@@ -147,13 +150,44 @@ internal static void CheckForDefaultUIActionMapChanges()
+                    else if (obj is InputActionAsset)
+                    {
+                        AssetDatabase.RemoveObjectFromAsset(obj);
+                    }
```
What happens if there is something here that isn't an InputActionReference or an InpuitAction asset?
I recommend we either throw or log an error or alternatively change the last else if into an unconditional else which would remove anything? Or is this designed to leave e.g. InputManager or any other root objects untouched?

lyndon-unity replied:
Its designed to leave InputManager and others untouched

ekcoh reviewed (approved):
Approving but have one outstanding questions around the intended behavior in the `DeleteActionAssetAndActionReferences()` function regarding intended behavior to consider before merge.

lyndon-unity committed:
Merge branch 'develop' into ISX_1839_reset_fix_to_remove_duplicates
